### PR TITLE
Add Efficient Attention support on ROCM

### PR DESCRIPTION
**Pull request by xinyazhang**
This patch implements `with sdpa_kernel(SDPBackend.EFFICIENT_ATTENTION):` by reusing AOTriton's accelerated SDPA implementation

Known limitations:
- Only supports MI200/MI300X GPUs
- Does not support varlen
- Does not support `CausalVariant`
- Optional arguments `causal_diagonal` and `seqlen_k` in `_efficient_attention_forward/backward` must be null
- Does not work well with inductor's SDPA rewriter. The rewriter has been updated to only use math and flash attention on ROCM.

This PR also uses a different approach of installing AOTriton binary instead of building it from source in the base docker image. More details on motivation: https://github.com/pytorch/pytorch/pull/124885#issuecomment-2153229129

`PYTORCH_TEST_WITH_ROCM=1 PYTORCH_TESTING_DEVICE_ONLY_FOR="cuda" python test/test_transformers.py` yields "55028 passed, 20784 skipped" results with this change.  [Previous result](https://hud.pytorch.org/pr/127528) of `test_transformers.py` was 0 error, 0 failure, 55229 skipped out of 75517 tests in total (the XML report does not contain total number of passed tests).

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @chauhang @d4l3k @jeffdaily @sunway513 @jithunnair-amd @pruthvistony @ROCmSupport @dllehr-amd @jataylo @hongxiayang @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire

### CODE REVIEW

**Comment by jithunnair-amd** [comment]
@malfet Please review. All ROCm CI jobs passed.

**Review by malfet — APPROVED** [review]
Check that it only affects ROCM parts without checking what it actually changes in the hip files

**Comment by xinyazhang** [comment]
I think I know why the UT failed on CUDA. See the commit message of https://github.com/pytorch/pytorch/pull/124885/commits/0f8257af71d8763b4621e41894b79617f0c8e112

**Comment by xinyazhang** [comment]
@pruthvistony do NOT rebase against viable/strict. It does not contain necessary commit that moves AOTriton to base docker image.

**Comment by jithunnair-amd** [comment]
@malfet Thanks for the approval! We are waiting for a green signal on the ROCm CI jobs (they timed out in the last attempt) before merging. We'll handle the merge once signal is green.

**Comment by xinyazhang** [comment]
> @xinyazhang The only ROCm CI job currently failing is https://github.com/pytorch/pytorch/actions/runs/9379259480/job/25848946879

`grep FAILED job-logs.txt` shows
```
2024-06-05T16:22:26.0905686Z inductor/test_compiled_optimizers.py::CompiledOptimizerParityTestsCUDA::test_correctness_Adadelta_use_closure_True_cuda_float32 FAILED [14.9466s] [ 98%]
2024-06-05T16:22:26.2341935Z FAILED [14.9466s] inductor/test_compiled_optimizers.py::CompiledOptimizerParityTestsCUDA::test_correctness_Adadelta_use_closure_True_cuda_float32 - torch._dynamo.exc.BackendCompilerFailed: backend='inductor' raised:
2024-06-05T16:34:56.7964726Z inductor/test_efficient_conv_bn_eval.py::EfficientConvBNEvalCudaTests::test_basic_cuda FAILED [47.7058s] [100%]
2024-06-05T16:34:56.8509340Z FAILED [47.7058s] inductor/test_efficient_conv_bn_eval.py::EfficientConvBNEvalCudaTests::test_basic_cuda - torch._dynamo.exc.BackendCompilerFailed: backend='inductor' raised:
2024-06-05T16:34:56.8528608Z inductor/test_efficient_conv_bn_eval.py::EfficientConvBNEvalCudaTests::test_basic_cuda FAILED [84.0695s] [100%]
2024-06-05T16:34:56.8889705Z FAILED [84.0695s] inductor/test_efficient_conv_bn_eval.py::EfficientConvBNEvalCudaTests::test_basic_cuda - torch._dynamo.exc.BackendCompilerFailed: backend='inductor' raised:
2024-06-05T16:34:56.8908006Z inductor/test_efficient_conv_bn_eval.py::EfficientConvBNEvalCudaTests::test_basic_cuda FAILED [0.9633s] [100%]
2024-06-05T16:34:56.9167623Z FAILED [0.9633s] inductor/test_efficient_conv_bn_eval.py::EfficientConvBNEvalCudaTests::test_basic_cuda - torch._dynamo.exc.BackendCompilerFailed: backend='inductor' raised:
```

Consider we have the `keep-going` label. This is the only failure, and it seems irrelevant to this PR.
(We may want to disable this UT, tho)

**Comment by jithunnair-amd** [comment]
Some observations regarding AOTriton build:
* Originally, this PR was continuing with the approach of building AOTriton from source in the base docker image 
* The build time for AOTriton on a linux.12x.large machine increased from [20 min](https://github.com/pytorch/pytorch/pull/127012#issue-2313836750) to [30 min](https://github.com/pytorch/pytorch/actions/runs/9379257816/job/25823929540)
* However, we found that in case PyTorch builds jobs are triggered in parallel with the docker builds jobs, the former will try to build the base docker image from scratch if the latter hasn't pushed the docker image yet. This means that the PyTorch build job will try to build aotriton on a linux.2x.large machine and time out at [4hrs](https://github.com/pytorch/pytorch/actions/runs/9379259480/job/25823912105)
* So we now install aotriton from a release tarball in the base docker image: `https://github.com/ROCm/aotriton/releases/download/${VER}/aotriton-${VER}-${MANYLINUX}_${ARCH}-${ROCMBASE}.tar.bz2`
* This reduces the base docker image build time from [1hr 44min](https://github.com/pytorch/pytorch/actions/runs/9379257816/job/25823929540) to [58 min](https://github.com/pytorch/pytorch/actions/runs/9404529885/job/25903670744)
* This also reduces the PyTorch build job duration (when building the base docker image from scratch) to [2hrs](https://github.com/pytorch/pytorch/actions/runs/9404531458/job/25903651472) (1hr 40 mins for building the base docker image on a linux.2x.large node)

**Comment by jithunnair-amd** [comment]
Adding @huydhn to review the changes wrt AOTriton build.
Adding @atalman as well since we plan to make these same changes in the wheel build flow as well by installing AOTriton from binary.

File: aten/src/ATen/native/transformers/cuda/attention.cu
Comment (drisspg) [inline aten/src/ATen/native/transformers/cuda/attention.cu:1075]: do you need this if the check is already in sdp_utils?

File: test/distributed/_tensor/test_attention.py
Comment (drisspg) [inline test/distributed/_tensor/test_attention.py:331]: is this another, "parametrize cant be empty" problem again?

File: test/inductor/test_fused_attention.py
Comment (drisspg) [inline test/inductor/test_fused_attention.py:320]: Is this going to cause issues for people using compile + rocm?

File: test/test_transformers.py
Comment (drisspg) [inline test/test_transformers.py:2579]: maybe we should extract this to something like, `platform supports long sequences`?

File: test/test_transformers.py
Comment (drisspg) [inline test/test_transformers.py:2598]: hmm but this actually is a counter point to the above?

File: test/test_transformers.py
Comment (drisspg) [inline test/test_transformers.py:2709]: out of curiosity what happens if you dont call this? will this oom?

**Review by drisspg — COMMENTED** [review]
left some overarching comments, nothing looks blocking though

File: aten/src/ATen/native/transformers/cuda/attention.cu
Comment (xinyazhang) [reply aten/src/ATen/native/transformers/cuda/attention.cu:1075]: Oh I'm following the practice of `mha_fwd`, which also perform the checks for GPU architecture.
https://github.com/pytorch/pytorch/blob/dc8e3c2e904f22f05646e1d87987459af8845d7b/aten/src/ATen/native/transformers/cuda/flash_attn/flash_api.cpp#L359-L363

I'm not so sure what will happen if SDPA is call on GPUs with different arches (e.g., MI200+Navi31). This check is the last line of defense to guarantee the target GPU does support the SDPA request.

File: test/distributed/_tensor/test_attention.py
Comment (xinyazhang) [reply test/distributed/_tensor/test_attention.py:331]: Passing `None` as `attention_fn` is a trick to ensure on every platform, `test_ring_attention_compile_attention_fn0` always refers to the test of FA, and `test_ring_attention_compile_attention_fn0` is always for ME.

(However I do agree the skipping message can be improved, but haven't found an elegant way)

File: test/test_transformers.py
Comment (xinyazhang) [reply test/test_transformers.py:2709]: Yes. This will reliably cause OOM on MI210 systems when testing with the whole suite (but cannot be reproduced with any individual test).

File: test/test_transformers.py
Comment (xinyazhang) [reply test/test_transformers.py]: Replaced with `MEM_EFF_CAPABILITY_MATCHES_SM80` in the latest commit

File: test/inductor/test_fused_attention.py
Comment (xinyazhang) [reply test/inductor/test_fused_attention.py]: We fixed this in commit https://github.com/pytorch/pytorch/pull/124885/commits/0ff88e2109ec2a4a2f2b5ced54c84b3a2dee9596, by forcing the rewriter to use `aten._scaled_dot_product_attention_math` on ROCM

**Pull request closed without merging**
